balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -955,3 +1220,94 @@ type ccStateSubscriber struct {
 func (c *ccStateSubscriber) OnMessage(msg any) {
 	c.transitions = append(c.transitions, msg.(connectivity.State))
 }
+
+// hangingServer is a server that accept a TCP connection and remains idle until
+// asked to close or respond to the connection. They can be used to control
+// how long it takes for a subchannel to report a TRANSIENT_FAILURE in tests.
+type hangingServer struct {
```

Comment:
We have a bunch of utilities in the `testutils` package that can be used I feel, instead of implementing this new hangingServer.

There is a blocking dialer that lets you block, fail or succeed connection attempts: https://github.com/grpc/grpc-go/blob/ec10e73f02a38fcf1e4f6e3cdee5f73c89bfa4ac/internal/testutils/blocking_context_dialer.go#L33
There is a restartable listener: https://github.com/grpc/grpc-go/blob/master/internal/testutils/restartable_listener.go#L37
There is a wrapped listener that returns a wrapped conn: https://github.com/grpc/grpc-go/blob/master/internal/testutils/wrappers.go

Reply:
`BlockingDialer` fits the requirement perfectly! I've replaced the `hangingServer` that I implemented with `BlockingDialer`.